src/DataStreams/IBlockOutputStream.h

```diff
@@ -42,6 +42,10 @@ class IBlockOutputStream : private boost::noncopyable
       */
     virtual void flush() {}
 
+    /** Final flush.
+      */
+    virtual void finalize() {}
```

Comment:
I tried to use `writeSuffix` for such purpose, but didn't reach the solution. Have you tried use `writeSuffix` ? 
Methods `finalize` and `writeSuffix` here looks overlapping, like it's something that we call once when all things done. Maybe it requires comment about what the key difference and why we need both.

Reply:
Hm, it seems my knowledge about writeSuffix was a little bit incorrect, that's why I've added one more method, OK, let's try writeSuffix instead.

Reply:
Hmm, `IOutputFormat` contains both `finalize` and `doWriteSuffix` that calls `finalize`
https://github.com/ClickHouse/ClickHouse/blob/adc103e33891908885ea44bd8e3245f54cc528f6/src/Processors/Formats/IOutputFormat.h#L80

@nikitamikhaylov says that it's better to use `IOutputFormat` instead of `IBlockOutputStream` because `IBlockOutputStream` like a legacy

Reply:
>IOutputFormat contains both finalize and doWriteSuffix that calls finalize

But why this should be a problem?

>@nikitamikhaylov says that it's better to use IOutputFormat instead of IBlockOutputStream because IBlockOutputStream like a legacy. But I have no idea how difficult it is.

Yeah, let's do this separately.

>So, maybe we can leave your first version in that case.

I don't want to introduce one more method for streams, I will try to make writeSuffix works (right now the problem is with double call of writeSuffix)

Reply:
Plus version with finalize() skips some streams (for example stream for writing to Memory engine) and it is hard to catch such errors (even if it can be done right now, what to do with new streams?)